src/build/_helpers.py

```diff
@@ -0,0 +1,104 @@
+import functools
+import os
+import subprocess
+import sys
+
+from typing import TYPE_CHECKING, AbstractSet, Any, Iterator, Mapping, Optional, Sequence, Tuple, Union
+
+
+ConfigSettingsType = Mapping[str, Union[str, Sequence[str]]]
+PathType = Union[str, 'os.PathLike[str]']
+
+if TYPE_CHECKING:
+    from typing_extensions import Literal, Protocol
+
+    Distribution = Literal['sdist', 'wheel']
+    WheelDistribution = Literal['wheel']
+
+    class RunnerType(Protocol):
+        def __call__(
+            self, cmd: Sequence[str], cwd: Optional[PathType] = None, env: Optional[Mapping[str, str]] = None
+        ) -> None:
+            """
+            Run a command in a Python subprocess.
+
+            The parameters mirror those of ``subprocess.run``.
+
+            :param cmd: The command to execute
+            :param cwd: The working directory
+            :param env: Variables to be exported to the environment
+            """
+
+
+def default_subprocess_runner(
+    cmd: Sequence[str], cwd: Optional[PathType] = None, env: Optional[Mapping[str, str]] = None
+) -> None:
+    subprocess.run(cmd, check=True, cwd=cwd, env=env)
+
+
+def quiet_subprocess_runner(
+    cmd: Sequence[str], cwd: Optional[PathType] = None, env: Optional[Mapping[str, str]] = None
+) -> None:
+    subprocess.run(cmd, check=True, cwd=cwd, env=env, stdout=subprocess.DEVNULL)
+
+
+def pep517ify_subprocess_runner(values: Union['RunnerType', Tuple['RunnerType', Optional[Mapping[str, str]]]]) -> Any:
```

Comment:
What does pep517ify means?

Reply:
Also any reason we did not annotate the return value correctly?

Reply:
It means to make the subprocess callback compatible with `pep517`. `pep517` expects there to be an `extra_environ` parameter but we need to be able to modify the environment, not just add values to it. I did not annotate the return value because `pep517` itself is not type hinted and mypy's support for function-altering decorator return types is not great and it would've probably involved one or several `cast`s or `type: ignore`s.

Reply:
> mypy's support for function-altering decorator return types is not great and it would've probably involved one or several `cast`s or `type: ignore`s.

Such as, I never had any problems with this :thinking: 

Perhaps rename it to pep517_pkg_ify, because currently I thought you're referring to something in PEP-517 and not the package interface we're using.

Reply:
Done. It was actually okay because it's not used as a decorator.